src/Parallel/ArrayCollection/SendDataToElement.hpp

```diff
@@ -0,0 +1,96 @@
+ * The number of messages needed to take the next time step on the receiver
+ * element is kept track of and a message is sent to the parallel runtime
+ * system (e.g. Charm++) only when the receiver/neighbor element has all the
+ * data it needs to take the next time step. This is done so as to reduce
+ * pressure on the runtime system by sending fewer messages.
+ */
+struct SendDataToElement {
+  using return_type = void;
+
+  template <typename ParallelComponent, typename DbTagList, size_t Dim,
+            typename ReceiveTag, typename ReceiveData, typename Metavariables>
+  static return_type apply(
+      db::DataBox<DbTagList>& box,
+      const gsl::not_null<Parallel::NodeLock*> /*node_lock*/,
+      const gsl::not_null<Parallel::GlobalCache<Metavariables>*> cache,
+      const ReceiveTag& /*meta*/, const ElementId<Dim>& element_to_execute_on,
+      typename ReceiveTag::temporal_id instance, ReceiveData&& receive_data) {
+    const size_t my_node = Parallel::my_node<size_t>(*cache);
+    auto& element = db::get_mutable_reference<
+                        typename ParallelComponent::element_collection_tag>(
+                        make_not_null(&box))
+                        .at(element_to_execute_on);
+    // While we don't mutate the value, we want to avoid locking the DataBox
+    // and the nodegroup by using `db::get_mutable_reference`. If/when we
+    // start dynamically inserting and removing elements, we'll need to update
+    // how we handle this. For example, we might need the containers to have
+    // strongly stability guarantees.
+    const size_t node_of_element =
+        db::get_mutable_reference<Parallel::Tags::ElementLocations<Dim>>(
+            make_not_null(&box))
+            .at(element_to_execute_on);
+    auto& my_proxy =
+        Parallel::get_parallel_component<ParallelComponent>(*cache);
+    if (node_of_element == my_node) {
+      [[maybe_unused]] size_t count = 0;
+      if constexpr (std::is_same_v<evolution::dg::AtomicInboxBoundaryData<Dim>,
+                                   typename ReceiveTag::type>) {
+        count = ReceiveTag::insert_into_inbox(
+            make_not_null(&tuples::get<ReceiveTag>(element.inboxes())),
+            instance, std::forward<ReceiveData>(receive_data));
+      } else {
+        // Scope so that we minimize how long we lock the inbox.
+        std::lock_guard inbox_lock(element.inbox_lock());
+        count = ReceiveTag::insert_into_inbox(
+            make_not_null(&tuples::get<ReceiveTag>(element.inboxes())),
+            instance, std::forward<ReceiveData>(receive_data));
+      }
+      // A lower bound for the number of neighbors is
+      // `2 * Dim - number_of_block_boundaries`, which doesn't give us the
+      // exact minimum number of sends we need to do, but gets us close in most
+      // cases. If we really wanted to we could also add the number of
+      // directions that don't have external boundaries in our neighbors block.
+      // if (count >=
```

Comment:
did you mean to leave this if commented out?

Reply:
Yes, because this is the bookkeeping code I talked with Will about to minimize communication. I want to get the rest in before I get the counting sorted out completely :)